woocommerce/woocommerce
# Include e2e, api and performance tests in ci.yml

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





A first iteration of adding support for all remaining test types in `ci-jobs` utility and `ci.yml` workflow.

Closes #41489  .

I tried to keep the existing functionality in place as much as possible with the least effort in this first iteration. The reporting can be better, we could use more abstraction in there so we won't need to duplicate jobs.

- Introduced test types (default, e2e, api, performance), for each of them a different list of jobs will be created. 
- Introduced `shardingArguments` configuration, that can be used to shard or split tests into multiple jobs. 
- Defined projects e2e, api, and performance (k6) tests for woocommerce plugin. The corresponding `pr-build-and-e2e-test.yml` and `smoke-test-pr-merge.yml` have been disabled (added a `workflow_dispatch` trigger to have a valid syntax). They should be removed in a follow up PR, I wanted to keep them for easier reference until we validate the new worklfow is what we need.

Out of scope of this PR - will be addressed in follow-ups:
- granular changes configuration.
- tests for other plugins (e.g. woocommerce-blocks e2e).
- editor performance tests (`metrics.yml`). While I intended to add them in this PR the setup for those tests requires more effort. 
- e2e tests with HPOS disabled - these have a separate workflow (`non-cot-pr-build-and-e2e-tests.yml`) and I'll leave them running as such for now.
- daily e2e tests and release e2e tests. While ci.yml will run for release events, the existing smoke-test-release.yml workflow is using a different env configuration. Same with the daily runs. Support for multiple environments will be part of a future iteration.

⚠️ When this PR gets merged the `Runs API tests.` and `Evaluate e2e tests results` have to be removed from the required checks.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Check the CI jobs and compare with the ci configuration defined in plugins/woocommerce/package.json.
The should be test jobs running as before, 5 e2e tests jobs running, one api tests job, one performance tests job.
The e2e test reports should be published in case of both failure and success.
There should be a Slack alert sent for e2e tests failures.
The Evaluate Project Job Statuses job should fail if at least one job of type lint, test, e2e, or api fails. It should not fail if performance tests are failing.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
